org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.xtend

```diff
@@ -101,18 +104,34 @@ class JavaDerivedStateComputer {
 		(resource as JavaResource).getCompilationUnit()
 	}
 	
+	protected def ClassFileCache findOrCreateClassFileCache(ResourceSet rs) {
+		var cache = ClassFileCache.findInEmfObject(rs)
+		if (cache === null) {
+			cache = new ClassFileCache
+			cache.attachToEmfObject(rs)
+		}
+		cache
+	}
+	
 	def void installFull(Resource resource) {
 		if (resource.isInfoFile)
 			return;
+		val classFileCache = resource.resourceSet.findOrCreateClassFileCache
 		val compilationUnit = getCompilationUnit(resource)
 		val classLoader = getClassLoader(resource)
 		
 		val data = resourceDescriptionsProvider.getResourceDescriptions(resource.resourceSet)
 		if (data === null)
 			throw new IllegalStateException("no index installed")
 		// TODO use container manager
-		val nameEnv = new IndexAwareNameEnvironment(resource, classLoader, data, stubGenerator)
+		val nameEnv = new IndexAwareNameEnvironment(resource, classLoader, data, stubGenerator, classFileCache)
 		val compiler = new Compiler(nameEnv, DefaultErrorHandlingPolicies.proceedWithAllProblems(), resource.compilerOptions, [
+			for (cls : it.classFiles) {
+				// TODO What is with inner classes (they contain $)
+				// TODO is there a better way to obtain the class name
+				val key = qualifiedNameConverter.toQualifiedName(new String(cls.fileName).replace("/","."))
```

Comment:
We probably do not need to dispatch through a QNConverter here. 
``` 
Splitter splitter = Splitter.on('/');
QualifiedName.create(splitter.splitToList(CharBuffer.wrap(fileName))
```
This would avoid repetitive string allocations and s&r operations.